src/transformer_util.ts

```diff
@@ -0,0 +1,545 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ts from 'typescript';
+import * as tsickle from './tsickle';
+
+/**
+ * Adjusts the given CustomTransformers with additional transformers
+ * to fix bugs in TypeScript.
+ * @param given A
+ */
+export function createCustomTransformers(given: ts.CustomTransformers): ts.CustomTransformers {
+  if (!given.after && !given.before) {
+    return given;
+  }
+  const before = given.before || [];
+  before.unshift(firstTransform);
+  before.push(preTypeScriptTransform);
+  const after = given.after || [];
+  after.unshift(postTypescriptTransform);
+  return {before, after};
+}
+
+/**
+ * Transform to be used as first transform to reset some state.
+ */
+function firstTransform(context: ts.TransformationContext) {
+  return (sourceFile: ts.SourceFile) => {
+    (context as TransformationContext).fileContext = new FileContext(sourceFile);
+    return sourceFile;
+  };
+}
+
+function assertFileContext(context: TransformationContext, sourceFile: ts.SourceFile): FileContext {
+  if (!context.fileContext) {
+    throw new Error(
+        `Illegal State: FileContext not initialized. ` +
+        `Did you forget to add the "firstTransform" as first transformer? ` +
+        `File: ${sourceFile.fileName}`);
+  }
+  if (context.fileContext.file.fileName !== sourceFile.fileName) {
+    throw new Error(
+        `Illegal State: File of the FileContext does not match. File: ${sourceFile.fileName}`);
+  }
+  return context.fileContext;
+}
+
+/**
+ * And extended version of the TransformationContext that stores the FileContext as well.
+ */
+interface TransformationContext extends ts.TransformationContext {
+  fileContext?: FileContext;
+}
+
+/**
+ * A context that stores information per file to e.g. allow communication
+ * between transformers.
+ * Note that the ts.TransformationContext stores context about all files
+ * in the emit.
+ */
+class FileContext {
+  syntheticNodeParents = new Map<ts.Node, ts.Node|undefined>();
+  importOrReexportDeclarations: Array<ts.Node&{moduleSpecifier: ts.StringLiteral}> = [];
+  lastCommentEnd = -1;
+  constructor(public file: ts.SourceFile) {}
+}
+
+/**
+ * Transform that needs to be executed right before TypeScript's transform.
+ *
+ * This prepares the node tree to workaround some bugs in the TypeScript emitter.
```

Comment:
Created https://github.com/angular/tsickle/issues/528 to track filing these bugs.